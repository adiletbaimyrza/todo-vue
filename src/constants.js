export const TodoFilter = Object.freeze({
  ALL: 'all',
  ACTIVE: 'active',
  COMPLETED: 'completed',
})
