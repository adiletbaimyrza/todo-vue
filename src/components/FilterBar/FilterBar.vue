<template>
  <div :class="$style['container']" :data-testid="DataTestIds.FILTER_BAR">
    <p
      :class="$style['container__count']"
      :data-testid="DataTestIds.ACTIVE_COUNT"
    >
      {{ activeCount }} items left
    </p>
    <ul
      :class="$style['container__filters']"
      :data-testid="DataTestIds.FILTERS"
    >
      <li>
        <button
          @click="setFilter(TodoFilter.ALL)"
          type="button"
          :class="{
            [$style['container__filters--active']]:
              currentFilter === TodoFilter.ALL,
          }"
          :data-testid="DataTestIds.FILTER_ALL"
        >
          all
        </button>
      </li>
      <li>
        <button
          @click="setFilter(TodoFilter.ACTIVE)"
          type="button"
          :class="{
            [$style['container__filters--active']]:
              currentFilter === TodoFilter.ACTIVE,
          }"
          :data-testid="DataTestIds.FILTER_ACTIVE"
        >
          active
        </button>
      </li>
      <li>
        <button
          @click="setFilter(TodoFilter.COMPLETED)"
          type="button"
          :class="{
            [$style['container__filters--active']]:
              currentFilter === TodoFilter.COMPLETED,
          }"
          :data-testid="DataTestIds.FILTER_COMPLETED"
        >
          completed
        </button>
      </li>
    </ul>
    <button
      @click="deleteCompleted"
      type="button"
      :class="$style['container__clear']"
      :data-testid="DataTestIds.CLEAR_COMPLETED"
    >
      clear completed
    </button>
  </div>
</template>

<script>
import { mapState, mapGetters, mapMutations } from 'vuex'
import { TodoFilter, DataTestIds } from '@/constants'

export default {
  name: 'FilterBar',
  computed: {
    ...mapState(['currentFilter']),
    ...mapGetters(['activeCount']),
  },
  methods: {
    ...mapMutations(['setFilter', 'deleteCompleted']),
  },
  data() {
    return {
      TodoFilter,
      DataTestIds,
    }
  },
}
</script>

<style lang="scss" module>
.container {
  display: flex;
  padding: 1em 1.5em;
  align-items: center;
  width: 600px;
  background-color: var(--card-background-color);

  &__count {
    font-size: 0.9em;
    color: #b1b1b1;

    &:hover {
      cursor: default;
    }
  }

  &__filters {
    flex-grow: 1;
    display: flex;
    justify-content: center;

    li button {
      color: #b1b1b1;
      font-family: Inter, sans-serif;
      font-weight: 500;
      padding: 0 1em;
      text-transform: capitalize;

      &:hover {
        color: var(--filter-text--hover);
      }
    }

    &--active {
      color: #1969ff !important;
    }
  }

  &__clear {
    font-family: Inter, sans-serif;
    text-transform: capitalize;
    color: #b1b1b1;

    &:hover {
      color: var(--filter-text--hover);
    }
  }
}
</style>
