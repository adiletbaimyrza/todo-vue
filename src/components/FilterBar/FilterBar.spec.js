import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'
import FilterBar from './FilterBar.vue'
import { TodoFilter } from '@/constants'

/* global describe beforeEach jest expect it */
describe('FilterBar.vue', () => {
  let store
  let state
  let getters
  let mutations
  let wrapper

  beforeEach(() => {
    state = {
      currentFilter: TodoFilter.ALL,
    }

    getters = {
      activeCount: () => 5,
    }

    mutations = {
      setFilter: jest.fn(),
      deleteCompleted: jest.fn(),
    }

    store = createStore({
      state,
      getters,
      mutations,
    })

    wrapper = shallowMount(FilterBar, {
      global: {
        plugins: [store],
      },
    })
  })

  it('renders correctly with initial state', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('calls setFilter mutation when filter button is clicked', async () => {
    const button = wrapper.find('button')
    await button.trigger('click')
    expect(mutations.setFilter).toHaveBeenCalled()
  })

  it('calls deleteCompleted mutation when clear button is clicked', async () => {
    const button = wrapper.find(`.container__clear`)
    await button.trigger('click')
    expect(mutations.deleteCompleted).toHaveBeenCalled()
  })
})
