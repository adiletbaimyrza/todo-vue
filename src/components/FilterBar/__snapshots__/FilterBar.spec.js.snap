// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar.vue renders correctly with initial state 1`] = `
<div
  class="container"
  data-testid="FilterBar"
>
  <p
    class="container__count"
    data-testid="ActiveCount"
  >
    5 items left 
  </p>
  <ul
    class="container__filters"
    data-testid="Filters"
  >
    <li>
      <button
        class="container__filters--active"
        data-testid="All"
        type="button"
      >
         all 
      </button>
    </li>
    <li>
      <button
        class=""
        data-testid="Active"
        type="button"
      >
         active 
      </button>
    </li>
    <li>
      <button
        class=""
        data-testid="Completed"
        type="button"
      >
         completed 
      </button>
    </li>
  </ul>
  <button
    class="container__clear"
    data-testid="Clear"
    type="button"
  >
     clear completed 
  </button>
</div>
`;
